SofaKernel/modules/SofaDefaultType/src/sofa/defaulttype/TypeInfoRegistry.cpp

```diff
@@ -0,0 +1,109 @@
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#include <typeindex>
+#include <map>
+#include <iostream>
+#include <sofa/helper/logging/Messaging.h>
+#include <sofa/defaulttype/AbstractTypeInfo.h>
+#include <sofa/defaulttype/DataTypeInfo.h>
+#include <sofa/defaulttype/typeinfo/NoTypeInfo.h>
+#include <sofa/helper/NameDecoder.h>
+#include <algorithm>
+#include "TypeInfoID.h"
+#include "TypeInfoRegistry.h"
+
+namespace sofa::defaulttype
+{
+
+static std::vector<const AbstractTypeInfo*>& getStorage()
+{
+    static std::vector<const AbstractTypeInfo*> typeinfos;
+    return typeinfos;
+}
+
+std::vector<const AbstractTypeInfo*> TypeInfoRegistry::GetRegisteredTypes(const std::string& target)
+{
+    bool selectAll = target == "";
+    std::vector<const AbstractTypeInfo*> tmp;
+    for(auto info : getStorage())
+    {
+        if(selectAll || info->getCompilationTarget() == target)
+            tmp.push_back(info);
+    }
+    return tmp;
+}
+
+const AbstractTypeInfo* TypeInfoRegistry::Get(const TypeInfoId& tid)
+{
+    sofa::Size id = tid.id;
+    auto& typeinfos = getStorage();
+
+    if( id < typeinfos.size() && typeinfos[id] != nullptr )
+        return typeinfos[id];
+
+    msg_error("TypeInfoRegistry") << "Missing type '"<< id << "' the type is not there..." << msgendl
+                                  << "     name: " << sofa::helper::NameDecoder::decodeFullName(tid.nfo);
+
+    return NoTypeInfo::get();
+}
+
+int TypeInfoRegistry::Set(const TypeInfoId& tid, AbstractTypeInfo* info, const std::string &compilationTarget)
+{
+    auto& typeinfos = getStorage();
+    sofa::Size id = tid.id;
+
+    if( info == nullptr )
+        return -1;
+
+    info->setCompilationTarget(compilationTarget);
+
+    if( id >= typeinfos.size() )
+    {
+        typeinfos.resize(id+1, NoTypeInfo::get());
+    }
+
+    if( typeinfos[id] != NoTypeInfo::get() )
+    {
+        if( typeinfos[id] != info && info->ValidInfo())
```

Comment:
not clear to me here how you could know if a new info is more valid than registered info (also flagged as valid)
So I would suggest a msg_warning rather than msg_info

Reply:
When we set a valid type info this override the other one only and only if it is a valid one (so we never override a typeinfo with one with a smaller level of validity()).